### PR TITLE
Fix NPE in ImageReference#isValidTag() and add more descriptive error messages

### LINKED ISSUE #1534: NullPointerException when a null additional tag is received
**Description of the issue**: A NullPointerException is thrown when an additional tag with null value is given.

**Expected behavior**: No exceptions

**Steps to reproduce**: Using jib Maven, reference an environment variable that is not provided.

**Environment**: Ubuntu 18.04, JDK 11.0.2, Maven 3.6.0, Jib Maven Plugin 1.0.2, Skaffold v0.24.0

**`jib-maven-plugin` Configuration**:
```xml
<configuration>
	<from>
		<image>adoptopenjdk/openjdk11:x86_64-alpine-jdk-11.0.2.9-slim</image>
	</from>
	<to>
		<image>company/project</image>
		<tags>
			<tag>${UNDEFINED}</tag>
		</tags>
	</to>
	<container>
		<user>nobody</user>
	</container>
</configuration>
```

**Log output**:
```
[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:1.0.2:dockerBuild (default) on project myproject: Execution default of goal com.google.cloud.tools:jib-maven-plugin:1.0.2:dockerBuild failed.: NullPointerException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.google.cloud.tools:jib-maven-plugin:1.0.2:dockerBuild (default) on project myproject: Execution default of goal com.google.cloud.tools:jib-maven-plugin:1.0.2:dockerBuild failed.
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default of goal com.google.cloud.tools:jib-maven-plugin:1.0.2:dockerBuild failed.
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:148)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: java.lang.NullPointerException
    at com.google.cloud.tools.jib.image.ImageReference.isValidTag (ImageReference.java:201)
    at com.google.cloud.tools.jib.api.Containerizer.withAdditionalTag (Containerizer.java:106)
    at java.lang.Iterable.forEach (Iterable.java:75)
    at com.google.cloud.tools.jib.plugins.common.PluginConfigurationProcessor.configureContainerizer (PluginConfigurationProcessor.java:461)
    at com.google.cloud.tools.jib.plugins.common.PluginConfigurationProcessor.processCommonConfiguration (PluginConfigurationProcessor.java:208)
    at com.google.cloud.tools.jib.plugins.common.PluginConfigurationProcessor.processCommonConfigurationForDockerDaemonImage (PluginConfigurationProcessor.java:79)
    at com.google.cloud.tools.jib.maven.BuildDockerMojo.execute (BuildDockerMojo.java:114)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
```

### PR DESCRIPTION
**Pull request by briandealwis**
Fixes #1534

### CODE REVIEW

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/ImageReference.java [hunk 1]
```diff
@@ -198,7 +198,7 @@ public static boolean isValidRepository(String repository) {
    * @return {@code true} if is a valid tag; {@code false} otherwise
    */
   public static boolean isValidTag(String tag) {
-    return tag.matches(TAG_REGEX) || tag.matches(DescriptorDigest.DIGEST_REGEX);
+    return tag != null && (tag.matches(TAG_REGEX) || tag.matches(DescriptorDigest.DIGEST_REGEX));
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/ImageReference.java]: `tag` isn't annotated with `@NotNull`, so checking null here doesn't seem pretty. Perhaps we should ensure `tag` is not null.

Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/image/ImageReference.java]: Oops, `@Nullable` I mean.

**Comment by TadCordle** [comment]
Should the build fail, or should we just warn+skip null tags?

**Comment by chanseokoh** [comment]
That's a good question. I think warn+skip seems reasonable. This can be handy if you a fixed `pom.xml` template while you want to skip additional tags when an env variable is not defined.

**Comment by briandealwis** [comment]
This doesn't seem much different from #416 / #1401 where we fail the build.

**Comment by chanseokoh** [comment]
Hmm... I was initially thinking CMD args is a bit different story because all the values in the list are related to form a single semantic CMD configuration. That is, there must be something wrong if any of the values are null, whereas in this case, each tag value is independent of each other so it's OK if some of them are null.

However, I am totally fine with failing for any null and have stronger consistency around config. I actually like it.

**Comment by briandealwis** [comment]
PTAL: Took a different tack and now validate the values provided from Maven.  Kept the helper message for `Containerizer.withAdditionalTag()` to provide useful information as to which tag was invalid.

File: jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenRawConfigurationTest.java [hunk 2]
```diff
@@ -105,4 +105,23 @@ public void testGetters() {
...
+
+  /**
+   * Maven turns empty tags into null entries (<a
+   * href="https://github.com/GoogleContainerTools/jib/issues/1534">#1534</a>).
+   */
+  @Test
+  public void testNullTags() {
+    JibPluginConfiguration jibPluginConfiguration = Mockito.mock(JibPluginConfiguration.class);
+    Mockito.when(jibPluginConfiguration.getTargetImageAdditionalTags())
+        .thenReturn(new HashSet<>(Arrays.asList(null)));
```
Comment (chanseokoh) [inline jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenRawConfigurationTest.java]: Guess `Arrays.asList(null)` throws NPE.

File: jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenRawConfigurationTest.java [hunk 3]
```diff
@@ -105,4 +105,23 @@ public void testGetters() {
...
+
+    MavenRawConfiguration rawConfiguration = new MavenRawConfiguration(jibPluginConfiguration);
+    try {
+      rawConfiguration.getToTags();
+      Assert.fail();
+    } catch (IllegalArgumentException ex) {
```
Comment (chanseokoh) [inline jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenRawConfigurationTest.java]: `Preconditions.checkNotNull` throws NPE, so this probably won't work.

File: jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java [hunk 4]
```diff
@@ -299,10 +299,10 @@ String getTargetImage() {
    */
   Set<String> getTargetImageAdditionalTags() {
     String property = getProperty(PropertyNames.TO_TAGS);
-    if (property != null) {
-      return new HashSet<>(ConfigurationPropertyValidator.parseListProperty(property));
-    }
-    return new HashSet<>(to.tags);
+    List<String> tags =
+        property != null ? ConfigurationPropertyValidator.parseListProperty(property) : to.tags;
+    tags.forEach(tag -> Preconditions.checkNotNull(tag, "null or empty tag"));
```
Comment (chanseokoh) [inline jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java]: I'd like to have the user see a more friendly and useful message. This is the Maven plugin code, so I guess it's OK to have a very Maven-specific message here. Or should we use `HelpfulSuggestions`?

File: jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenRawConfiguration.java [hunk 5]
```diff
@@ -69,7 +71,9 @@ public AuthProperty getToAuth() {
 
   @Override
   public Iterable<String> getToTags() {
-    return jibPluginConfiguration.getTargetImageAdditionalTags();
+    Set<String> tags = jibPluginConfiguration.getTargetImageAdditionalTags();
```
Comment (chanseokoh) [inline jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenRawConfiguration.java]: I actually preferred doing this check in `JibPluginConfiguration`. I intended `MavenRawConfiguration` to be a super-thin adapter between `RawConfiguration` and `JibPluginConfiguration` without any processing or validation. Doing the validation in `JibPluginConfiguration` may be weird as well, but given the state of `MavenRawConfiguration`, at least we can keep `MavenRawConfiguration` untainted.

That said, I've been thinking config value validation is a mess currently. It's being done here and there haphazardly for different configs.

Comment (briandealwis) [reply jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenRawConfiguration.java]: I agree that `JibPluginConfiguration` is the better place: it also knows whether the value came from the `jib.to.tags`  system property or `<jib><to><tags>` element.  I started it there but testing became messy.

File: jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java [hunk 6]
```diff
@@ -299,10 +300,13 @@ String getTargetImage() {
    */
   Set<String> getTargetImageAdditionalTags() {
     String property = getProperty(PropertyNames.TO_TAGS);
-    if (property != null) {
-      return new HashSet<>(ConfigurationPropertyValidator.parseListProperty(property));
-    }
-    return new HashSet<>(to.tags);
+    List<String> tags =
+        property != null ? ConfigurationPropertyValidator.parseListProperty(property) : to.tags;
+    String source = property != null ? PropertyNames.TO_TAGS : "<jib><to><tags>";
```
Comment (TadCordle) [inline jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java]: ~~`<jib>`~~`<to><tags`

Comment (chanseokoh) [reply jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java]: Oh, yeah, good catch.

**Pull request merged**